### ISSUE TITLE
Kubectl returns pods in various states when only those in Running state are selected via --field-selector, or -o jsonpath

### ISSUE THREAD

**Issue by wknapik** [kubernetes/kubectl#450]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):
No.

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):
field-selector, field selector, status phase

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.1", GitCommit:"d4ab47518836c750f9949b9e0d387f20fb92260b", GitTreeState:"clean", BuildDate:"2018-04-12T14:26:04Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.7", GitCommit:"b30876a5539f09684ff9fde266fda10b37738c9c", GitTreeState:"clean", BuildDate:"2018-01-16T21:52:38Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**:
AWS
- **OS** (e.g. from /etc/os-release):
Arch Linux
- **Kernel** (e.g. `uname -a`):
Linux something 4.16.3-1-ARCH #1 SMP PREEMPT Thu Apr 19 09:17:56 UTC 2018 x86_64 GNU/Linux
- **Install tools**:
?
- **Others**:
?

**What happened**:
```
kubectl get pods --field-selector=status.phase=Running --namespace some-ns
```
returns pods in other states than Running.

**What you expected to happen**:
I expected only pods in state Running to be returned.

**How to reproduce it** (as minimally and precisely as possible):
```
kubectl get pods --field-selector=status.phase=Running --namespace some-ns
```

**Anything else we need to know**:
That's way too deep... I don't think I can tackle that question here...

**Comment by wknapik** [comment]
Selecting pods like this
```
kubectl get pods --namespace some-ns -o jsonpath='{.items[?(@.status.phase=="Running")]}'
```
also doesn't work as expected. Pods in states like Terminating and ContainerCreating are returned.
In fact, if you run
```
kubectl get pods --namespace some-ns -o jsonpath='{.items[?(@.status.phase=="Running")].status.phase}'
```
you'll see everything described as Running, even though
```
kubectl get pods --namespace some-ns
```
will show the same pods in other states.

I feel like I'm missing some key piece of information here, but the cheatsheet here https://kubernetes.io/docs/reference/kubectl/cheatsheet/ actually says
```
# Get all running pods in the namespace
$ kubectl get pods --field-selector=status.phase=Running
```

**Comment by pashau** [comment]
we have the same problem

**Comment by wknapik** [comment]
@seans3 I'd say that's a bug, not a feature request. The code doesn't do what the documentation says it would. And it doesn't error out, it returns incorrect results.

**Comment by mackoftrack** [comment]
+1 

I'm seeing this bug. 

```
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.3", GitCommit:"a4529464e4629c21224b3d52edfe0ea91b072862", GitTreeState:"clean", BuildDate:"2018-09-10T11:44:36Z", GoVersion:"go1.11", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.7-gke.6", GitCommit:"06898a4d0f2b96f82b43d9e59fa2825bd3d616a2", GitTreeState:"clean", BuildDate:"2018-10-02T17:32:01Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by DovAmir** [comment]
+1

**Comment by DovAmir** [comment]
happens also on

> Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T21:04:45Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"darwin/amd64"}
> Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.4", GitCommit:"bf9a868e8ea3d3a8fa53cbb22f566771b3f8068b", GitTreeState:"clean", BuildDate:"2018-10-25T19:06:30Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}

**Comment by DovAmir** [comment]
btw, this is not a selector error, 
for some reason, a pod in a terminating state has 
> status:
> ...phase: Running	

in its yaml

By looking at the yaml, the only way to identify that a pod is terminating, 
 is by the addition of the following 2 fields in metadata.
It is not possible to use selectors on these 2 fields

> metadata:
> ...deletionGracePeriodSeconds: 30
> ...deletionTimestamp:'2018-12-11T15:42:14Z'

my workaround hack to getting running pods that are NOT terminating is with the following command :

kubectl get pod | grep Running

**Comment by vithati** [comment]
This looks like not a bug. ```kubectl get pods``` output ```STATUS``` column not shows status.phase. ```kubectl get pods``` displays [table format](https://github.com/kubernetes/kubernetes/blob/master/pkg/printers/internalversion/printers.go#L580) of ```PodList``` object and uses ```status.containerStatuses``` states to display ```STATUS```  column data.

Pod phase valid states are [pending, running, succeeded, failed and unknown](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase). From this [document](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#example-states) 'Running' state depends on 'restartPolicy'.

Jsonpath applies on ```'PodList'``` object, not on kubectl output.  ```status.containerStatuses``` gives whether pod containers running or not.   Field selectors vary by Kubernetes resource types. status.containerStatuses not supported in pods field selectors.

**Comment by vithati** [comment]
--field-selector, or -o jsonpath depends on object data, not on kubectl get table format output. I think adding more information to command comment helps to resolve this issue. 

In https://kubernetes.io/docs/reference/kubectl/cheatsheet/ 
'#Get all running pods in the namespace' change to '#Get all running pods in the namespace and Running state depends on 'restartPolicy'. Refer https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#example-states for pod status.phase states'

**Comment by idealhack** [comment]
/close

feel free to reopen if it still an issue

**Comment by wknapik** [comment]
I see no PRs referencing this issue, which means it's still valid. Why was this closed ?

**Comment by soltysh** [comment]
Closing per previous comment.

**Comment by wknapik** [comment]
@soltysh which previous comment ?

Nothing has changed since I reported this, afaik. `kubectl` still reports something else than the user is expecting (and what the documentation is saying it would) and there's still no way to get pods in state `Running` without scraping unstructured output. I don't understand why this was closed.

**Comment by howardjohn** [comment]
I think this should be open again.

The [docs](https://kubernetes.io/docs/reference/kubectl/cheatsheet/) clearly state

> Get all running pods in the namespace
> $ kubectl get pods --field-selector=status.phase=Running

This command doesn't get all pods in running state.

Call it a bug, a doc error, or even a UX issue of confusing command, but something seems missing here. 

I should be able to query for running pods without doing some hack like `kubectl get pods | grep Running`

**Comment by sassanh** [comment]
Currently there's no easy way to select running pods (as running means in English language)
So please reopen this issue and do not close it until API supports this feature as it seems to be a basic feature to list running pods.

This is my dirty workaround for now: `kubectl get pods  -o jsonpath='{.items[*].status.containerStatuses[*].state.running},{.items[*].metadata.name}' --field-selector=status.phase==Running | sed 's/ /\n/' | grep startedAt | awk -F',' '{print $2}'`

I'm sure you agree this is not how it's supposed to be.

**Comment by vmrm** [comment]
`# Get all running pods in the namespace
kubectl get pods --field-selector=status.phase=Running` 
from https://kubernetes.io/docs/reference/kubectl/cheatsheet/#interacting-with-running-pods

**Comment by gdankov** [comment]
Can you please reopen this issue?

This is still not working as documented!

**Comment by wknapik** [comment]
@idealhack 
> feel free to reopen if it still an issue

This should never have been closed. The issue is still there and, iuuc, nothing's been done to address it. Please reopen.

**Comment by idealhack** [comment]
/reopen

but I'll defer more discussion to @kubernetes/sig-cli-bugs

**Comment by alahijani** [comment]
As a workaround for now I'm relying on the `READY` column showing `n/n`:

```
kubectl get pod | awk -F' *|/' '$2 == $3 && $4 == "Running"'
```

**Comment by brianpursley** [comment]
I looked into this issue a little bit today and here is what I found out.

First of all, for reference, here is how I am reproducing:
```
kubectl run nginx --image=nginx && sleep 1 && kubectl delete deploy nginx && sleep 1 && kubectl get pods --field-selector=status.phase=Running
```
The output is:
```
deployment.apps/nginx created
deployment.apps "nginx" deleted
NAME                     READY   STATUS        RESTARTS   AGE
nginx-6db489d4b7-k4k4j   1/1     Terminating   0          2s
```

At least by my understanding, this is the problem:  I told kubectl only to get pods with `--field-selector=status.phase=Running` but I see the pod in a Terminating status.

The reason this happens is that the value displayed in the table's status column is constructed on-the-fly based on many different factors.  It is *not* the actual status.phase of the pod, which contributes to it, but there are other factors as well.

In this situation, the code reports a status of "Terminating" because when the pod is being deleted, the DeletionTimestamp != nil.  See code excerpt below to see where this occurs:

https://github.com/kubernetes/kubernetes/blob/fe2fdcd6953a877664cbdb784888c0e7f636ce40/pkg/printers/internalversion/printers.go#L780
```
        if pod.DeletionTimestamp != nil && pod.Status.Reason == node.NodeUnreachablePodReason {
		reason = "Unknown"
	} else if pod.DeletionTimestamp != nil {
		reason = "Terminating"
	}
```

**Technically** the pod is still in the Running status phase even though Terminating is displayed as the Status in the table.  So the field selector is technically correct, although it is confusing.

---

So given that this issue has been open for a long time, I am trying to understand a couple of things:

1.  **Is this still an issue where the code needs to be changed, and if so, what is the proposed new behavior?**  I think this issue is stuck because it is not clear what needs to change.

2.  **Could this be addressed without a code change?  For example, by improving the documentation** to explain the difference between the displayable status and status.phase can be explained?

Feel free to respond, anyone who has more information or an opinion on this issue.

**Comment by wknapik** [comment]
@brianpursley getting running pods from a script should be easy and reliable. Currently it's not. You could change the documentation, but that would not help users with that problem.

And yes, the current documentation is actively misleading and should be changed, but that's not where the work should stop.

IMO.

**Comment by brianpursley** [comment]
Thanks @wknapik 

By "running pods" do you mean you do not want to see the ones that are still running, but in the process of terminating?  Because that is what it is showing you now.  When it says Terminating in the `kubectl get` output, that is technically still a pod that is in a Running status phase, it just shows you "Terminating" because it knows it is in the process of terminating, but that is not an actual status.

I'm not trying to be difficult, just trying to define exactly what the expected behavior should be so that the issue can potentially be worked on.

I see one of the [workarounds above](https://github.com/kubernetes/kubectl/issues/450#issuecomment-571791955) is to use awk to examine the Ready column.  Is it that you are really interested in knowing how many pods are completely ready?  Ready being defined as 100% of the containers in the pod have ready == true.

**Comment by brianpursley** [comment]
I was curious to see if there was a way using only kubectl to get a list of the ready pods, and this is what I came up with (inspired by [alahijani's awk workaround above](https://github.com/kubernetes/kubectl/issues/450#issuecomment-571791955))  

It uses a [go-template](https://golang.org/pkg/text/template/) to print the name of pods only if all of its containers are ready.

It's definitely more complex than the awk workaround, but if you are looking for a pure kubectl way, or want to extend it with other conditions or fields, this go-template approach might be an option.

```
kubectl get pods -o=go-template --template='{{range .items}}{{$ready:=true}}{{range .status.containerStatuses}}{{if not .ready}}{{$ready = false}}{{end}}{{end}}{{if $ready}}{{.metadata.name}}{{"\n"}}{{end}}{{end}}'
```

**Comment by vosmax** [comment]
I've simplified @brianpursley solution a bit.

Find NOT READY pods: 
`kubectl -n default get pods -o=go-template --template='{{range $i := .items}}{{range .status.containerStatuses}}{{if not .ready}}{{$i.metadata.name}}{{"\n"}}{{end}}{{end}}{{end}}'`

Find READY pods: 
 `kubectl -n default get pods -o=go-template --template='{{range $i := .items}}{{range .status.containerStatuses}}{{if .ready}}{{$i.metadata.name}}{{"\n"}}{{end}}{{end}}{{end}}'`

**Comment by GMZwinge** [comment]
@brianpursley and @vosmax: the kubectl commands using the go-template don't work in my case. They handle the pods with a `STATUS` of `Terminating` as if they had a `STATUS` of `Running`.

**Comment by vosmax** [comment]
@GMZwinge this is normal behavior. While termination is going, container has status `Running`, because it got termination signal and trying kindly complete the main thread. When it's terminated pod changes his `phase` on state `Terminated`.  
[Here](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/) you may find answers on your questions.

**Comment by DirkTheDaring** [comment]
Get running pods and do not show Terminating pods (.metata.deletiion != nil) 

`kubectl get pods --field-selector=status.phase=Running --template {{range .items}}{{ if not .metadata.deletionTimestamp }}{{.metadata.name}}{{"\n"}}{{end}}{{end}}`

**Comment by lallinger-arbeit** [comment]
I understand the issue with people wanting to get pods which are in terminating state, but at least there should be a flag to add to tell kubectl it really should only output pods which also show running state when calling `kubectl get pods`.

Thanks @DirkTheDaring, your solution outputs the desired information!

**Comment by szpak** [comment]
It would still be nice to have it possible without any "templating magic".

/remove-lifecycle rotten

**Comment by danielyaa5** [comment]
Ya this is hella dumb, status says terminating not running... lets not overcomplicate things or at least update documentation

**Comment by danielyaa5** [comment]
```
 ➜ kubectl get pods --field-selector=status.phase=Running --template {{range .items}}{{ if not .metadata.deletionTimestamp }}{{.metadata.name}}{{"\n"}}{{end}}{{end}}
error: name cannot be provided when a selector is specified
```

**Comment by danielyaa5** [comment]
This did work for me though ```
kubectl get pods --template '{{range .items}}{{ if not .metadata.deletionTimestamp }}{{.metadata.name}}{{"\n"}}{{end}}{{end}}'```

**Comment by jakerobb** [comment]
/remove-lifecycle rotten

This is still an issue. A pod should have an easily selectable field with values that map intuitively to the values `kubectl get pods` shows in the Status column: Running, Terminated, ContainerCreating, CrashLoopBackOff, ImagePullBackOff, etc. Ideally, `kubectl` should then have a standard option (preferably not just --field-selector) to let you show only pods with (or without) a specific status. 

Thanks for the workaround, @DirkTheDaring!

**Comment by RinkiyaKeDad** [comment]
Unassigning myself since I haven't gotten a chance to look at this and probably won't have the bandwidth for this for some time 🙂

/unassign

**Comment by szpak** [comment]
It would still be nice to have an easy way to achieve that.

/remove-lifecycle stale

**Comment by eddiezane** [comment]
This still needs to be answered before we can proceed.

> By "running pods" do you mean you do not want to see the ones that are still running, but in the process of terminating? Because that is what it is showing you now. When it says Terminating in the kubectl get output, that is technically still a pod that is in a Running status phase, it just shows you "Terminating" because it knows it is in the process of terminating, but that is not an actual status.

Showing the pod as `Terminating` is a convenience here. As Brian said the pod is indeed still in a state of `Running` but has a `metadata.deletionTimestamp` meaning the pod is marked for deletion.

Do you want the the convenience removed? This would show the same pods but with the state column of `Running` even if they are marked for deletion.

Do you want it to say `Running (Terminating)`?

**Comment by PowerSurge1** [comment]
> `metadata.deletionTimestamp` meaning the pod is marked for deletion.

This is certainly what I was after. Thanks!

**Comment by guai** [comment]
tl;dr
so, what is a right way to recreate a deployment and `wait` for the pod to become ready excluding those pesky terminating pods?

**Comment by wknapik** [comment]
@guai you can try https://gist.github.com/wknapik/315f52a20a2af43da35be5412625794a. I haven't used it in a long time, so ymmv, but it used to work fine for me. There may be simpler/better ways, but I'll leave those for someone else to cover ;]
